src/core/graph.js

```diff
@@ -19,88 +21,175 @@ export type Edge<+T> = {|
-  _nodes: AddressMap<Node<NP>>;
-  _edges: AddressMap<Edge<EP>>;
+type MaybeNode<+NP> = {|+address: Address, +node: Node<NP> | void|};
 
-  // The keyset of each of the following fields should equal the keyset
-  // of `_nodes`. If `e` is an edge from `u` to `v`, then `e.address`
-  // should appear exactly once in `_outEdges[u.address]` and exactly
-  // once in `_inEdges[v.address]` (and every entry in `_inEdges` and
-  // `_outEdges` should be of this form).
-  _outEdges: AddressMap<{|+address: Address, +edges: Address[]|}>;
-  _inEdges: AddressMap<{|+address: Address, +edges: Address[]|}>;
+export class Graph<NP, EP> {
+  // Invariant: sizes of `_nodeIndices`, `_nodes`, `_outEdges`, and
+  // `_inEdges` are all equal.
+  _nodeIndices: AddressMap<{|+address: Address, +index: Integer|}>;
+  _nodes: MaybeNode<NP>[];
+  _edges: AddressMap<IndexedEdge<EP>>;
+  _outEdges: Address[][];
+  _inEdges: Address[][];
 
   constructor() {
-    this._nodes = new AddressMap();
+    this._nodeIndices = new AddressMap();
+    this._nodes = [];
     this._edges = new AddressMap();
-    this._outEdges = new AddressMap();
-    this._inEdges = new AddressMap();
+    this._outEdges = [];
+    this._inEdges = [];
   }
 
   copy(): Graph<$Supertype<NP>, $Supertype<EP>> {
     return Graph.mergeConservative(new Graph(), this);
   }
 
   equals(that: Graph<NP, EP>): boolean {
-    return this._nodes.equals(that._nodes) && this._edges.equals(that._edges);
+    const theseNodes = this.nodes();
+    const thoseNodes = that.nodes();
+    if (theseNodes.length !== thoseNodes.length) {
+      return false;
+    }
+    for (const node of theseNodes) {
+      if (!deepEqual(node, that.node(node.address))) {
+        return false;
+      }
+    }
+
+    const theseEdges = this.edges();
+    const thoseEdges = that.edges();
+    if (theseEdges.length !== thoseEdges.length) {
+      return false;
+    }
+    for (const edge of theseEdges) {
+      if (!deepEqual(edge, that.edge(edge.address))) {
+        return false;
+      }
+    }
+
+    return true;
   }
 
-  toJSON(): GraphJSON<NP, EP> {
+  toJSON(): Compatible<GraphJSON<NP, EP>> {
+    const partialNodes: {|
+      key: string,
+      oldIndex: Integer,
+      data: {|
+        +address: Address,
+        +payload?: NP,
+      |},
+    |}[] = this._nodes
+      .map((maybeNode, oldIndex) => {
+        const key = stringify(maybeNode.address);
+        const data = maybeNode.node || {address: maybeNode.address};
+        return {key, oldIndex, data};
+      })
+      .filter(({oldIndex: idx}) => {
+        // Eliminate dangling nodes with neither in- nor out-edges:
+        // e.g., nodes that exist as a result of an `addNode` followed
+        // by a `removeNode` and no additional incident edges.
+        return (
+          this._nodes[idx].node !== undefined ||
+          this._outEdges[idx].length > 0 ||
+          this._inEdges[idx].length > 0
+        );
+      });
+    partialNodes.sort((a, b) => {
```

Comment:
Do we have a unit test verifying that the JSON representation is independent of node insertion order?
I know we test this for equality; we should check for JSON equality too.